compliance/collection/main.go

```diff
@@ -182,17 +217,106 @@ func manageNodeScanLoop(ctx context.Context, rescanInterval time.Duration, scann
 	return sensorC
 }
 
-func scanNode(nodeName string, scanner nodeinventorizer.NodeInventorizer) (*sensor.MsgFromCompliance, error) {
-	result, err := scanner.Scan(nodeName)
+func runCachedScan(opts *cachedScanOpts) (*storage.NodeInventory, error) {
+	inventory, err := opts.scanner.Scan(opts.nodeName)
 	if err != nil {
 		return nil, err
 	}
+	inv, err := proto.Marshal(inventory)
+	if err != nil {
+		return nil, err
+	}
+	if err := os.WriteFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath), inv, 0600); err != nil {
+		return nil, err
+	}
+
+	return inventory, nil
+}
+
+func createAndObserveMessage(nodeName string, inventory *storage.NodeInventory) *sensor.MsgFromCompliance {
 	msg := &sensor.MsgFromCompliance{
 		Node: nodeName,
-		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: result},
+		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: inventory},
 	}
 	cmetrics.ObserveInventoryProtobufMessage(msg)
-	return msg, nil
+	return msg
+}
+
+// scanNodeWithBackoff runs scans with a linear backoff based on a file to not overstrain a Node if the container keeps restarting.
+// The backoff file will only be encountered if the previous container is killed during a call to scanNodeWithBackoff.
+// Note: This does not prevent strain in case of repeated pod recreation, as it is based on an EmptyDir.
+func scanNodeWithBackoff(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	backoffIntervalSeconds := int64(env.NodeInventoryInitialBackoff.DurationSetting() / time.Second)
+
+	backoffFile, err := os.ReadFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath))
```

Comment:
Do you mean `backoffFileContents` instead of `backofFile`? I don't have a better name for this variable, but my point is that somethingFile usually has a very specific meaning of being a file descriptor.

Reply:
Done